icefoganalytics/sfa-client
# Refactor: Simplify and Genericise Letter Generation Pipeline

Fixes https://github.com/icefoganalytics/sfa-client/issues/46
Depends on https://github.com/icefoganalytics/sfa-client/pull/45

# Context

The original letter generation pipeline was built as a best guess of what we would need. Now its out of date and hard to read.
This project is a quest to simply the code to a point where it is easy to add new templates.

# Implementation

Adds the following endpoints:
- `/api/v2/admin/funding-requests/:fundingRequestId/letters/:letterSlug` - supports .json, .html, or .pdf format
- `/api/v2/admin/funding-requests/:fundingRequestId/letters` - lists possible letter types for a given funding request
- `/api/v2/admin/funding-request-letters` - lists all possible funding request letters

Letter generation pipeline does
1. Has a controller that load the letter builder service.
2. The letter builder service selects the correct letter service for a given template
3. Letter service retrieves the letter template path and serializes the data as required for the specific template. It uses a per-template serializer for this purpose.


# Testing Instructions

1. Boot the api and database services via `dev up`
4. Boot the web service via `cd src/web && npm run start`
5. Log in to the app at http://localhost:8080/
6. Go to http://localhost:3000/api/v2/admin/funding-request-letters to list all the letter metadata
7. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters to list the letter options for a given request
8. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters/institution.html to preview the institution letter. 
    > Letter will crash as dataset is not sufficient to build letter.
9. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters/institution.pdf to download the institution letter as a pdf.
    > Letter will crash as dataset is not sufficient to build letter.
10. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters/institution.json to debug why the letter is crashing.
11. Check that http://localhost:3000/api/v2/admin/funding-requests/221/letters/institution.html returns a 422 because there is no letter handler for cancelled requests.
12. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters/student.html to preview the student letter.
    > Letter will crash as dataset is not sufficient to build letter.
10. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters/student.pdf to download the student letter as a pdf.
    > Letter will crash as dataset is not sufficient to build letter.
